flake-extension/src/main/java/com/palantir/flake/FlakeRetryExtension.java

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * An {@link Extension} that retries methods annotated with the {@link FlakeRetryTest} annotation. This class shouldn't
+ * be extended by test methods. It should be used via the {@link FlakeRetryTest} annotation.
+ * <p>
+ * Like other JUnit Extensions, this extension respects lifecycle of test methods. Thus, methods annotated
+ * {@link BeforeEach} will run before *each* attempt at running a test, and methods annotated
+ * with {@link AfterEach} will similarly run after each attempt. However, since retrying takes place at the method level,
+ * methods annotated with {@link BeforeAll} and {@link AfterAll} will NOT be re-run.
+ * <p>
+ * To illustrate this, consider a test class with one flaky method that has been annotated with {@link FlakeRetryTest}.
+ * The order in which methods are run is as follows:
+ * <pre>
+ *     BeforeAll runs
+ *     BeforeEach runs
+ *     Test fails on attempt 1
+ *     AfterEach runs
+ *     BeforeEach runs
+ *     Test succeeds on attempt 2
+ *     AfterEach runs
+ *     AfterAll runs
+ * </pre>
+ * This extension is intended to catch flaky cases in the test method itself. It will not catch flaky implementations of
+ * {@link BeforeAll}, {@link BeforeEach}, {@link AfterAll}, {@link AfterEach}.
+ * <p>
+ * Note: Please be very careful about ordering when chaining this with other JUnit {@link Extension}s.
+ */
+final class FlakeRetryExtension
+        implements TestTemplateInvocationContextProvider, AfterTestExecutionCallback, TestExecutionExceptionHandler {
+
+    private static final SafeLogger log = SafeLoggerFactory.get(FlakeRetryExtension.class);
+
+    private int maxNumberOfRetriesUntilSuccess;
+    private List<Class<? extends Throwable>> retryableExceptions;
+
+    private int numberOfRetriesSoFar = 1; // This is increased after each iteration. This is why it is initialized to 1
+    private boolean hasSucceeded = false;
+    private boolean hasCompletelyFailed = false;
+
+    @Override
+    public boolean supportsTestTemplate(ExtensionContext extensionContext) {
+        return isAnnotated(extensionContext.getTestMethod(), FlakeRetryTest.class);
+    }
+
+    @Override
+    public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(
+            ExtensionContext extensionContext) {
+        FlakeRetryTest testWithFlakeRetry = findAnnotation(extensionContext.getTestMethod(), FlakeRetryTest.class)
+                .orElseThrow(() -> new IllegalStateException(
+                        "Expected to find @TestWithFlakeRetry annotation on the test method, but couldn't find."
+                                + " This is likely to be a problem with " + FlakeRetryExtension.class.getName()
+                                + " implementation."));
+
+        retryableExceptions = List.of(testWithFlakeRetry.retryableExceptions());
+        maxNumberOfRetriesUntilSuccess = testWithFlakeRetry.maxNumberOfRetriesUntilSuccess();
+        Preconditions.checkArgument(
```

Comment:
I'm surprised error prone hasn't complained here (and it doesn't really matter here, in all honesty), but prefer using Preconditions from 'com.palantir.safe-logging:preconditions'